src/controls/QskTextEdit.h

```diff
@@ -0,0 +1,202 @@
+/******************************************************************************
+ * QSkinny - Copyright (C) The authors
+ *           SPDX-License-Identifier: BSD-3-Clause
+ *****************************************************************************/
+
+#ifndef QSK_TEXT_EDIT_H
+#define QSK_TEXT_EDIT_H
+
+#include "QskControl.h"
+#include "QskTextOptions.h"
+
+class QValidator;
+class QskFontRole;
+
+class QSK_EXPORT QskTextEdit : public QskControl
+{
+    Q_OBJECT
+
+    Q_PROPERTY( QString text READ text
+        WRITE setText NOTIFY textChanged USER true )
+
+    Q_PROPERTY( QskFontRole fontRole READ fontRole
+        WRITE setFontRole RESET resetFontRole NOTIFY fontRoleChanged )
+
+    Q_PROPERTY( QFont font READ font )
+
+    Q_PROPERTY( Qt::Alignment alignment READ alignment
+        WRITE setAlignment RESET resetAlignment NOTIFY alignmentChanged )
+
+    Q_PROPERTY( QskTextOptions::WrapMode wrapMode READ wrapMode
+        WRITE setWrapMode NOTIFY wrapModeChanged )
+
+    Q_PROPERTY( int lineCount READ lineCount NOTIFY lineCountChanged )
+
+    Q_PROPERTY( QskTextOptions::TextFormat textFormat READ textFormat
+        WRITE setTextFormat NOTIFY textFormatChanged )
+
+    Q_PROPERTY( ActivationModes activationModes READ activationModes
+        WRITE setActivationModes NOTIFY activationModesChanged )
+
+    Q_PROPERTY( bool selectByMouse READ selectByMouse
+        WRITE setSelectByMouse NOTIFY selectByMouseChanged )
+
+    Q_PROPERTY( bool editing READ isEditing
+        WRITE setEditing NOTIFY editingChanged )
+
+    Q_PROPERTY( qreal tabStopDistance READ tabStopDistance
+        WRITE setTabStopDistance NOTIFY tabStopDistanceChanged )
+
+    using Inherited = QskControl;
+
+  public:
+    QSK_SUBCONTROLS( TextPanel, Text )
+    QSK_STATES( ReadOnly, Editing, Error )
+
+    enum ActivationMode
+    {
+        NoActivation,
+
+        ActivationOnFocus = 1 << 0,
+        ActivationOnMouse = 1 << 1,
+        ActivationOnKey = 1 << 2,
+
+        ActivationOnInput = ActivationOnMouse | ActivationOnKey,
+        ActivationOnAll = ActivationOnFocus | ActivationOnMouse | ActivationOnKey
+    };
+
+    Q_ENUM( ActivationMode )
+    Q_DECLARE_FLAGS( ActivationModes, ActivationMode )
+
+    enum EchoMode
```

Comment:
do we need this mode here ?

Reply:
Right, that was copied over from QskTextInput, I just removed it.